rmw_connext_shared_cpp/src/security_logging.cpp

```diff
@@ -0,0 +1,140 @@
+const char log_verbosity_variable_name[] = "ROS_SECURITY_LOG_VERBOSITY";
+
+// Connext property names. These are being taken from table 10.1 of
+// https://community.rti.com/static/documentation/connext-dds/5.3.1/doc/manuals/connext_dds/dds_security/RTI_SecurityPlugins_GettingStarted.pdf
+const char log_file_property_name[] = "com.rti.serv.secure.logging.log_file";
+const char distribute_enable_property_name[] = "com.rti.serv.secure.logging.distribute.enable";
+const char verbosity_property_name[] = "com.rti.serv.secure.logging.log_level";
+
+const struct
+{
+  const char * const name;
+  const char * const level;
+} supported_verbosities[] =
+{
+  {"EMERGENCY", "0"},
+  {"ALERT", "1"},
+  {"CRITICAL", "2"},
+  {"ERROR", "3"},
+  {"WARNING", "4"},
+  {"NOTICE", "5"},
+  {"INFORMATIONAL", "6"},
+  {"DEBUG", "7"},
+};
+
+bool string_to_verbosity_level(const char * str, const char ** level)
+{
+  for (const auto & item : supported_verbosities) {
+    if (strcmp(str, item.name) == 0) {
+      *level = item.level;
+      return true;
+    }
+  }
+
+  return false;
+}
+
+bool apply_property(
+  DDS::PropertyQosPolicy & policy, const char * const property_name,
+  const char * const value)
+{
+  // Overwrite existing properties, so remove it if it already exists
+  DDS::PropertyQosPolicyHelper::remove_property(policy, property_name);
+
+  auto status = DDS::PropertyQosPolicyHelper::add_property(
+    policy,
+    property_name,
+    value,
+    DDS::BOOLEAN_FALSE);
+
+  return status == DDS::RETCODE_OK;
+}
+
+bool get_env(const char * variable_name, const char ** variable_value)
+{
+  const char * error_message = rcutils_get_env(variable_name, variable_value);
+  if (error_message != NULL) {
+    RMW_SET_ERROR_MSG_WITH_FORMAT_STRING(
+      "unable to get %s environment variable: %s",
+      variable_name,
+      error_message);
+    return false;
+  }
+
+  return true;
+}
+}  // namespace
+
+rmw_ret_t apply_security_logging_configuration(DDS::PropertyQosPolicy & policy)
+{
+  const char * env_value;
+
+  // Handle logging to file
+  if (!get_env(log_file_variable_name, &env_value)) {
+    return RMW_RET_ERROR;
+  }
+  if (strlen(env_value) > 0) {
+    if (!apply_property(policy, log_file_property_name, env_value)) {
+      RMW_SET_ERROR_MSG("failed to set security logging file");
+      return RMW_RET_ERROR;
+    }
+  }
+
+  // Handle log distribution over DDS
+  if (!get_env(log_publish_variable_name, &env_value)) {
+    return RMW_RET_ERROR;
+  }
+  if (strlen(env_value) > 0) {
+    if (!apply_property(policy, distribute_enable_property_name, env_value)) {
+      RMW_SET_ERROR_MSG("failed to set security logging distribute");
+      return RMW_RET_ERROR;
+    }
+  }
+
+  // Handle log verbosity
+  if (!get_env(log_verbosity_variable_name, &env_value)) {
+    return RMW_RET_ERROR;
+  }
+  if (strlen(env_value) > 0) {
+    const char * level;
+    if (!string_to_verbosity_level(env_value, &level)) {
```

Comment:
Both `log_file` and `log_publish` values are not checked for correctness (e.g. is a bool) and this is, I assume, delayed to Connext. But here you are checking that the value is within the enum. How about returning instead a default error-level and let Connext do the checking as well for consistency?

Reply:
Good point. I'd prefer to be consistent the other direction though and just enforce true/false here. I'm not sure what connext supports, but different vendors may support different things. We should limit that to what we know is supported.

Reply:
> We should limit this to what we know is supported across them.

Which might be vastly different `true/True/on/On/1/One/tRuE/verdadera` and whatnot. Furthermore, I'd argue that this is a wrapper, let keep it thin and not duplicate any logic.

Reply:
> Which might be vastly different true/True/on/On/1/One/tRuE/verdadera

Right, that's exactly my point. Ideally one would be able to set these environment variables (or define the XML that will follow) not caring what RMW is being used. If we don't limit what we support here the possible values will diverge depending on RMW.

Reply:
gotcha :+1:

Reply:
Alright I added this validation for booleans. The file path still seems reasonable to be handled by the DDS vendor, not least because dealing with file paths in a cross platform way here seems error-prone.